open-hackathon-server/src/hackathon/hazure/azure_formation.py

```diff
@@ -0,0 +1,697 @@
+
+        # next step: setup storage
+        self.scheduler.add_once("azure_formation", "setup_storage", context=sctx, seconds=0)
+        return
+
+    def setup_storage(self, sctx):
+        # get context from super context
+        ctx = sctx.job_ctxs[sctx.current_job_index]
+        adapter = StorageAccountAdapter(ctx.subscription_id, ctx.pem_url, host=ctx.management_host)
+
+        try:
+            if not adapter.storage_account_exists(ctx.storage_account_name):
+                # TODO: use the async way
+                if not adapter.create_storage_account(
+                        ctx.storage_account_name,
+                        ctx.storage_account_description,
+                        ctx.storage_account_label,
+                        ctx.storage_account_location):
+                    self.log.error("azure virtual environment %d create storage account failed via creation" %
+                                   sctx.current_job_index)
+                    self.__on_setup_failed(sctx)
+                    return
+
+                # create storage account remote successfully, record
+                sctx.remote_created.append(Context(
+                    type="storage_account",
+                    name=ctx.storage_account_name))
+        except Exception as e:
+            self.log.error(
+                "azure virtual environment %d create storage account failed: %r"
+                % (sctx.current_job_index, str(e)))
+            self.__on_setup_failed(sctx)
+            return
+
+        self.log.debug("azure virtual environment %d storage setup done" % sctx.current_job_index)
+
+        # next step: setup virtual machine
+        self.scheduler.add_once("azure_formation", "setup_virtual_machine", context=sctx, seconds=0)
+
+    def setup_virtual_machine(self, sctx):
+        # get context from super context
+        ctx = sctx.job_ctxs[sctx.current_job_index]
+
+        adapter = VirtualMachineAdapter(ctx.subscription_id, ctx.pem_url, host=ctx.management_host)
+
+        if adapter.deployment_exists(ctx.cloud_service_name, ctx.deployment_slot):
+            # TODO: need to store the deployment info into db?
+            self.__setup_virtual_machine_with_deployment_existed(sctx)
+        else:
+            self.__setup_virtual_machine_without_deployment_existed(sctx)
+
+    def __setup_virtual_machine_with_deployment_existed(self, sctx):
+        # get context from super context
+        ctx = sctx.job_ctxs[sctx.current_job_index]
+
+        adapter = VirtualMachineAdapter(ctx.subscription_id, ctx.pem_url, host=ctx.management_host)
+
+        deployment_name = adapter.get_deployment_name(ctx.cloud_service_name, ctx.deployment_slot)
+        if not ctx.is_vm_image:
+            network_config = get_network_config(
+                ctx.raw_network_config,
+                adapter.get_assigned_endpoints(ctx.cloud_service_name))
+        else:
+            network_config = None
```

Comment:
Are both branches are tested? I remember that behavior of `OS` and `VM` are different on azure